### PR TITLE
Expose some useful tlv macros

### PR DESCRIPTION
**Pull request by mariocynicys**
This exposes the 2 macros `impl_writeable_msg!` & `impl_writeable!` and any macros/structs called/used inside them.
Also all the imports (`use`s) in the macros are prefixed with `$crate` (essentially, making them absolute paths) so the macro works fine with any rust edition (2015 and others).

I have added some docs for the `export`ed macros & `pub`ed structs based on my understanding of them.

Let me know if I should expose more/less macros or edit docs.

Addressing #1539

### CODE REVIEW

File: lightning/src/util/ser_macros.rs [hunk 1]
```diff
@@ -254,10 +293,13 @@ macro_rules! impl_writeable_msg {
 	}
 }
 
+/// Implements Readable/Writeable for a struct. Note that this macro doesn't handle messages
+/// containing TLV records. If your message contains TLVs, use `impl_writeable_msg!` instead.
+#[macro_export]
```
[the remark this answers is not in the document]
Comment (meryacine) [reply lightning/src/util/ser_macros.rs:298]: hmmm, why should they do it as tlv??
A custom message (types `32768`-`65535`) may have a non-tlv fields.

File: lightning/src/util/ser_macros.rs [hunk 2]
```diff
@@ -226,25 +242,48 @@ macro_rules! decode_tlv_stream {
...
+/// lightning::impl_writeable_msg!(LightningMessage, {
+/// 	to,
+/// 	note,
+/// 	secret_number,
+/// }, {
+/// 	(1, nick_name, option),
+/// 	(3, street_number, option),
+/// });
+/// ```
+#[macro_export]
```
[the remark this answers is not in the document]
Comment (meryacine) [reply lightning/src/util/ser_macros.rs:272]: Fine.
So what macros should be exposed??
I'm thinking `{decode,encoder}_tlv_stream` (for non varint prefixed uses) and `{read,write}_tlv_fields` (for varint prefixed uses).

**Comment by tnull** [comment]
@meryacine Are you still interested in continuing this PR? Would be great to have the serialization macros exported to be used downstream!

**Comment by meryacine** [comment]
@tnull Yeah sure. I ended up copying what I needed from these macros and modified them a bit that's why the PR got staled.

> Would be great to have the serialization macros exported to be used downstream.

Are we talking about `{decode,encoder}_tlv_stream` here?

**Comment by tnull** [comment]
> Are we talking about `{decode,encoder}_tlv_stream` here?

Well, I'm currently implementing serialization manually, so the macros would mostly help brevity and readability. It would therefore be great to have `impl_writeable_tlv_based`.  As it is however dependent on quite a few other macros, we'd probably need to expose them too. We should however as @TheBlueMatt mentioned take some precautions, i.e., mark them as internal  with an `_` prefix and document their usage.

**Comment by tnull** [comment]
This needs a rebase now.

**Comment by tnull** [comment]
Closing as superseded by #1823.

**Pull request closed without merging**